lib/pure/collections/sequtils.nim

```diff
@@ -401,25 +401,30 @@ template foldr*(sequence, operation: expr): expr =
     result = operation
   result
 
-template mapIt*(seq1, typ, op: expr): expr =
+template mapIt*[T,S](seq1: openArray[T], op: expr): seq[S] =
   ## Convenience template around the ``map`` proc to reduce typing.
   ##
   ## The template injects the ``it`` variable which you can use directly in an
-  ## expression. You also need to pass as `typ` the type of the expression,
-  ## since the new returned sequence can have a different type than the
-  ## original.  Example:
+  ## expression. Example:
   ##
   ## .. code-block::
   ##   let
   ##     nums = @[1, 2, 3, 4]
-  ##     strings = nums.mapIt(string, $(4 * it))
+  ##     strings = nums.mapIt($(4 * it))
   ##   assert strings == @["4", "8", "12", "16"]
-  var result {.gensym.}: seq[typ] = @[]
+  type T {.gensym.} = type(seq1[0])
+  type S {.gensym.} = type((
+    block:
+      var it{.inject.}: T
+      op))
+  var result {.gensym.} = newSeq[S](len(seq1))
+  var pos = 0
   for it {.inject.} in items(seq1):
-    result.add(op)
+    result[pos] = op
+    inc pos
   result
 
-template mapIt*(varSeq, op: expr) =
+template replaceIt*[T](varSeq: var seq[T], op: expr) =
```

Comment:
Just to clarify, we are talking about the "in place" map and mapIt here.

Since I removed the unnecessary extra type parameter from the other (ordinary) mapIt, the two mapIt templates have the same number of parameters, and the compiler fails to use the right one.